### DISCUSSION TITLE
Confusion about conflicting expressions and precedence (in C#)

### DISCUSSION THREAD

**Discussion by tamasvajk** [tree-sitter/tree-sitter#2024] [Q&A]
The current version of the C# language [specifies](https://github.com/tree-sitter/tree-sitter-c-sharp/blob/3b661ced8d510aa424e2f68314520c281dd8c113/grammar.js#LL115C5-L115C45) the conflict `[$.assignment_expression, $._expression]`. If I remove it, this is the message I get from tree-sitter:
```
❯ tree-sitter generate
Unresolved conflict for symbol sequence:

  '*'  _lvalue_expression  •  '='  …

Possible interpretations:

  1:  '*'  (_expression  _lvalue_expression)  •  '='  …
  2:  '*'  (assignment_expression  _lvalue_expression  •  assignment_operator  _expression)  (precedence: 0, associativity: Right)

Possible resolutions:

  1:  Specify a higher precedence in `assignment_expression` than in the other rules.
  2:  Specify a higher precedence in `_expression` than in the other rules.
  3:  Specify a left or right associativity in `_expression`
  4:  Add a conflict for these rules: `assignment_expression`, `_expression`
```

Why is there a conflict? Why isn't `'*'  _lvalue_expression` reduced to `_pointer_indirection_expression` which has higher precedence (17) than `assignment_expression`?

This question was originally asked in https://github.com/tree-sitter/tree-sitter-c-sharp/issues/273 while investigating recent size increases of `src/parser.c`.

**Comment by ahelwer** [comment]
Figuring out why there is a conflict usually requires in-depth understanding of both the syntax of the specific language and how the specific grammar is written. If I'm *really* confused I debug these by deleting all the irrelevant rules until I have a minimal grammar that still reproduces the issue. I haven't yet run into a spurious conflict but occasionally I have seen that my static precedence annotations weren't taking effect because they were made at the wrong level in the grammar, either at the parent rule of where they should have been or the child rule of where they should have been. Try experimenting with that, maybe.

**Reply by tamasvajk** [reply]
Thanks @ahelwer for the answer. I also don't have a better investigation approach than to start deleting rules from the grammar, I wanted to avoid that, because C# grammar is rather large, so it's a cumbersome way of pinpointing the issue. I was hoping that someone with in-depth knowledge on precedence rules and conflicts would be able to give better pointers to what might cause the issue here.

Regarding precedence definitions on wrong levels: I don't fully understand your point. Do you have a case in mind? The above "Possible interpretations" section shows that assignment has precedence 0, and the conflicting symbol sequence is:
```
'*'  _lvalue_expression  •  '='  …
```
I think the main question is that where can `'*'  _lvalue_expression '='` show up in a parse tree. And I'm under the impression that it can only be in 
- `_pointer_indirection_expression`, which is higher precedence and should be reduced
- `binary_expression` in the form `A * B =`
- variable declaration `int* p =`

But removing all of these doesn't solve the above conflict.
